runtime/src/in_mem_accounts_index.rs

```diff
@@ -648,34 +683,38 @@ impl<T: IndexValue> InMemAccountsIndex<T> {
 
         // this becomes inclusive - that is ok - we are just roughly holding a range of items.
         // inclusive is bigger than exclusive so we may hold 1 extra item worst case
-        let inclusive_range = Some(start..=end);
+        let inclusive_range = start..=end;
         let mut ranges = self.cache_ranges_held.write().unwrap();
+        let mut already_held = false;
         if start_holding {
-            ranges.push(inclusive_range);
+            if only_add_if_already_held {
+                for r in ranges.iter() {
+                    if r.contains(&start) && r.contains(&end) {
+                        already_held = true;
+                        break;
+                    }
+                }
+            }
+            if already_held || !only_add_if_already_held {
+                ranges.push(inclusive_range);
+            }
         } else {
             // find the matching range and delete it since we don't want to hold it anymore
-            let none = inclusive_range.is_none();
-            for (i, r) in ranges.iter().enumerate() {
-                if r.is_none() != none {
-                    continue;
-                }
-                if !none {
-                    // neither are none, so check values
-                    if let (Bound::Included(start_found), Bound::Included(end_found)) = r
-                        .as_ref()
-                        .map(|r| (r.start_bound(), r.end_bound()))
-                        .unwrap()
-                    {
-                        if start_found != &start || end_found != &end {
-                            continue;
-                        }
+            // search backwards, assuming LIFO ordering
+            for (i, r) in ranges.iter().enumerate().rev() {
+                if let (Bound::Included(start_found), Bound::Included(end_found)) =
+                    (r.start_bound(), r.end_bound())
+                {
+                    if start_found != &start || end_found != &end {
+                        continue;
                     }
                 }
                 // found a match. There may be dups, that's ok, we expect another call to remove the dup.
                 ranges.remove(i);
                 break;
```

Comment:
So if the `if let` is false, then that counts as a match? Is that for fully open ranges? If not, maybe it's cleaner to replace the if block at line 708 to be:

```rust
if start_found == &start && end_found == &end {
    ranges.remove(i);
    break;
}
```